core/config/config_helper.hpp

```diff
@@ -0,0 +1,220 @@
+// SPDX-FileCopyrightText: 2017 - 2024 The Ginkgo authors
+//
+// SPDX-License-Identifier: BSD-3-Clause
+
+#ifndef GKO_CORE_CONFIG_CONFIG_HELPER_HPP_
+#define GKO_CORE_CONFIG_CONFIG_HELPER_HPP_
+
+
+#include <string>
+#include <type_traits>
+
+
+#include <ginkgo/core/base/exception_helpers.hpp>
+#include <ginkgo/core/base/lin_op.hpp>
+#include <ginkgo/core/base/math.hpp>
+#include <ginkgo/core/config/config.hpp>
+#include <ginkgo/core/config/registry.hpp>
+#include <ginkgo/core/solver/solver_base.hpp>
+#include <ginkgo/core/stop/criterion.hpp>
+
+
+namespace gko {
+namespace config {
+
+
+/**
+ * LinOpFactoryType enum is to avoid forward declartion, linopfactory header,
+ * two template versions of build_from_config
+ */
+enum class LinOpFactoryType : int { Cg = 0 };
+
+
+/**
+ * It is only an intermediate step after dispatching the class base type. Each
+ * implementation needs to deal with the template selection.
+ */
+template <LinOpFactoryType flag>
+deferred_factory_parameter<gko::LinOpFactory> build_from_config(
```

Comment:
These functions doesn't seem necessary (or rather they should be renamed). Instead of the flag we could just have `cg_build_from_config` for example. These functions are stored in a lookup table anyway, so there already exists a mapping from `string -> function`.

Reply:
Yes, but I think it will make creating the function map much easier.
with specific function name, we also need to forward declation all function and repeated the function signature everytime.

Reply:
To create the function map you still have to explicitly write the flag. This is no different from using a different function name.

Reply:
Yes, but the repeated thing is less